cardano-db-sync/src/Cardano/DbSync/Epoch.hs

```diff
@@ -70,89 +61,270 @@ epochInsert trce (BlockDetails cblk details) = do
+  case cacheEpoch of
+    -- Technically there will always be something in the cache once this function is called
+    -- but to be exhaustive this is here.
+    Nothing -> do
+      liftIO . logInfo trce $ "replaceEpoch: CacheEpoch is Nothing so using DB to calculate epoch "
+      calculateEpochUsingDBQuery cache slotEpochNum epochId
+    Just cEpoch -> do
+      case ceEpoch cEpoch of
+        Nothing -> do
+          liftIO . logInfo trce $ "replaceEpoch: ceEpoch in cache is Nothing "
+          tryWithEpochFromDB cache cEpoch ePlutusPrices slotEpochNum epochId slotDetails
+          -- calculateEpochUsingDBQuery cache slotEpochNum epochId
+        Just ceE -> do
+          let newCalculatedEpoch = calculateEpochUsingCache ePlutusPrices cEpoch ceE slotDetails
+          liftIO . logInfo trce $ "replaceEpoch: calculated epoch from cache: " <> textShow newCalculatedEpoch
+          void $ writeEpochToCacheEpoch cache newCalculatedEpoch
+          Right <$> replace epochId newCalculatedEpoch
+
+calculateEpochUsingDBQuery ::
+  MonadIO m =>
+  Cache ->
+  Word64 ->
+  DB.EpochId ->
+  ReaderT SqlBackend m (Either SyncNodeError ())
+calculateEpochUsingDBQuery cache slotEpochNum epochId = do
+  -- this query is an expensive lookup so should be rarely used
+  newEpoch <- DB.queryCalcEpochEntry slotEpochNum
+  -- add the newly calculated epoch to cache.
+  void $ writeEpochToCacheEpoch cache newEpoch
+  -- replace the current epoch in the DB with our newly calculated epoch
+  Right <$> replace epochId newEpoch
+
+tryWithEpochFromDB ::
+  MonadIO m =>
+  Cache ->
+  CacheEpoch ->
+  EpochPlutusAndPrices ->
+  Word64 ->
+  DB.EpochId ->
+  SlotDetails ->
+  ReaderT SqlBackend m (Either SyncNodeError ())
+tryWithEpochFromDB cache ceE ePlutusPrices slotEpochNum epochId slotDetails = do
+  -- get the last known epoch from DB as there isn't one is CacheEpoch
+
+  -- need to check if this is a rollback because getting the last epoch from DB would create incorrect information.
+  latestEpochFromDb <- DB.queryLatestEpoch
+  case latestEpochFromDb of
+    -- if there isn't one in the DB as well then this is going to be the first block
+    Nothing -> do
+      calculateEpochUsingDBQuery cache slotEpochNum epochId
+    Just ep -> do
+      let newCalculatedEpoch = calculateEpochUsingCache ePlutusPrices ceE ep slotDetails
+      void $ writeEpochToCacheEpoch cache newCalculatedEpoch
+      Right <$> replace epochId newCalculatedEpoch
+
+
+
+calculateEpochUsingCache ::
+  EpochPlutusAndPrices ->
+  CacheEpoch ->
+  DB.Epoch ->
+  SlotDetails ->
+  DB.Epoch
+calculateEpochUsingCache ePlutusPrices cacheEpoch ceE slotDetails = do
+  let eIsEplutus = epochIsPlutusExtra ePlutusPrices
+      ePrices = epochPrices ePlutusPrices
+  case ceBlock cacheEpoch of
+    BlockByron blk ->
+      calculateEpochByron
+        blk
+        ceE
+        cacheEpoch
+        slotDetails
+    BlockShelley blk ->
+      calculateEpochGeneric
+        (Generic.fromShelleyBlock blk)
```

Comment:
I think it isn't necessary to redo the calculations of the Generic Block using `fromShelleyBlock`, `fromAllegraBlock` etc and take cases bassed on the era. This is already done when the block is received. Any needed parts of the block (txs count, output sum etc) can be passed similarly to the fees. In fact I'm not sure it's necessary to pass the whole blk.